internal/daemon/logstashconfig/file.go

```diff
@@ -0,0 +1,154 @@
+package logstashconfig
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/pkg/errors"
+
+	config "github.com/breml/logstash-config"
+	"github.com/breml/logstash-config/ast"
+	"github.com/breml/logstash-config/ast/astutil"
+
+	"github.com/magnusbaeck/logstash-filter-verifier/v2/internal/daemon/idgen"
+)
+
+type File struct {
+	Name string
+	Body []byte
+
+	config *ast.Config
+}
+
+func (f File) Save(targetDir string) error {
+	err := os.MkdirAll(path.Join(targetDir, path.Dir(f.Name)), 0700)
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(path.Join(targetDir, f.Name), f.Body, 0600)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (f *File) parse() error {
+	if f.config != nil {
+		return nil
+	}
+
+	icfg, err := config.Parse(f.Name, f.Body)
+	if err != nil {
+		return err
+	}
+
+	cfg, ok := icfg.(ast.Config)
+	if !ok {
+		return errors.New("not a valid logstash config")
+	}
+
+	f.config = &cfg
+
+	return nil
+}
+
+func (f *File) ReplaceInputs() error {
+	err := f.parse()
+	if err != nil {
+		return err
+	}
+
+	for i := range f.config.Input {
+		astutil.ApplyPlugins(f.config.Input[i].BranchOrPlugins, replaceInputs)
+	}
+
+	f.Body = []byte(f.config.String())
+
+	return nil
+}
+
+func replaceInputs(c *astutil.Cursor) {
+	if c.Plugin() == nil || c.Plugin().Name() == "pipeline" {
+		return
+	}
+
+	// TODO: __lfv_input must reflect the actual input, that has been replaced, such that this input
+	// can be referenced in the test case configuration.
+	c.Replace(ast.NewPlugin("pipeline", ast.NewStringAttribute("address", "__lfv_input", ast.Bareword)))
+}
+
+func (f *File) ReplaceOutputs() ([]string, error) {
+	err := f.parse()
+	if err != nil {
+		return nil, err
+	}
+
+	outputs := outputPipelineReplacer{
+		outputs: make([]string, 0),
+	}
+	for i := range f.config.Output {
+		astutil.ApplyPlugins(f.config.Output[i].BranchOrPlugins, outputs.walk)
+	}
+
+	f.Body = []byte(f.config.String())
+
+	return outputs.outputs, nil
+}
+
+type outputPipelineReplacer struct {
+	outputs []string
+}
+
+func (o *outputPipelineReplacer) walk(c *astutil.Cursor) {
+	if c.Plugin() == nil || c.Plugin().Name() == "pipeline" {
+		return
+	}
+
+	id, err := c.Plugin().ID()
+	if err != nil {
+		id = idgen.New()
+	}
+	outputName := fmt.Sprintf("lfv_output_%s", id)
+	o.outputs = append(o.outputs, id)
+
+	c.Replace(ast.NewPlugin("pipeline", ast.NewArrayAttribute("send_to", ast.NewStringAttribute("", outputName, ast.DoubleQuoted))))
+}
+
+func (f *File) Validate() error {
+	err := f.parse()
+	if err != nil {
+		return err
+	}
+
+	v := validator{}
+
+	for i := range f.config.Input {
+		astutil.ApplyPlugins(f.config.Input[i].BranchOrPlugins, v.walk)
+	}
+	for i := range f.config.Filter {
+		astutil.ApplyPlugins(f.config.Filter[i].BranchOrPlugins, v.walk)
+	}
+	for i := range f.config.Output {
+		astutil.ApplyPlugins(f.config.Output[i].BranchOrPlugins, v.walk)
+	}
+
+	if len(v.noIDs) > 0 {
+		return errors.Errorf("no IDs found for %v", v.noIDs)
```

Comment:
This is fixed in my branch, where I continued to implement (https://github.com/breml/logstash-filter-verifier/commit/5f5f034baa4a35abbb679f32dfd2052d847315b2#diff-e011f0d7266b9caea0022d59013bd493a6d9be725a6d1f922116a32891c3c401R140).

The IDs are necessary for the following cases:

1. Track coverage by querying Logstash for the statistics. Without ID, we can not know which filters have bin hit and which did not (not yet implemented).
2. Mock a filter
3. The input and output plugins are linked by the ID for each new test run (that is, the input pipeline with the generator sends the events to the input plugin of the system under test. for the outputs it works similarily).

Maybe we can come up with a small tool based on the improved https://github.com/breml/logstash-config. The new ast walk features in https://pkg.go.dev/github.com/breml/logstash-config@v0.4.0/ast/astutil would make this pretty simple.
An idea could be to generate the ID from the plugin type (e.g. `mutate` and an index, something like `mutate_1`). The tool would just add an ID to every plugin such that the user can update these IDs later.

Adding the ID in LFV without storing them in the logstash config files does not make much sense, because if you have a filter, that is not covered by the tests, you will have a hard time finding it without the ID.

Reply:
Yeah, I feel that some kind of tool support is necessary to not create a significant barrier for migration.